include/igl/tetrahedron_tetrahedron_adjacency.cpp

```diff
@@ -0,0 +1,134 @@
+// This file is part of libigl, a simple c++ geometry processing library.
+//
+// Copyright (C) 2017 Zhongshi Jiang <zhongshi@cims.nyu.edu>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/.
+#include "tetrahedron_tetrahedron_adjacency.h"
+#include "all_edges.h"
+#include "unique_simplices.h"
+#include "parallel_for.h"
+#include "unique_edge_map.h"
+#include "slice.h"
+#include <Eigen/Core>
+#include <algorithm>
+#include <iostream>
+#include <array>
+namespace igl
+{
+  struct tetrahedron_tetrahedron_adjacency_tet_cell
+  {
+    int v1;
+    int v2;
+    int v3;
+    int t;
+    int f;
+  };
+
+// Extract the face adjacencies
+  template<typename DerivedT, typename TTT_type, typename DerivedTT,
+           typename DerivedTTif, typename DerivedTTie>
+  IGL_INLINE void tetrahedron_tetrahedron_adjacency_extractTTi(
+      const Eigen::MatrixBase<DerivedT> &F,
+      std::vector<TTT_type> &TTT,
+      Eigen::MatrixBase<DerivedTT> &TT,
```

Comment:
The convention in libigl is to use MatrixBase for input types and PlainObjectBase for output types.

This avoids the need for `.derived()` calls and also prevents outputting into places that can't be resized properly.